rx_signal_processing/dsp.py

```diff
@@ -0,0 +1,369 @@
+
+        """
+        # [num_slices, num_beams, num_antennas]
+        beam_phases = np.array(beam_phases)
+
+        # [num_slices, num_beams, num_samples]
+        final_shape = (filtered_samples.shape[0], beam_phases.shape[1], filtered_samples.shape[2])
+        final_size = np.dtype(np.complex64).itemsize * reduce(lambda a,b: a*b, final_shape)
+        bf_shm = shared_memory.SharedMemory(create=True, size=final_size)
+        self.beamformed_samples = np.ndarray(final_shape, dtype=np.complex64, buffer=bf_shm.buf)
+        self.beamformed_samples[...] = np.einsum('ijk,ilj->ilk', filtered_samples, beam_phases)
+
+        self.shared_mem['bfiq'] = bf_shm
+
+    @staticmethod
+    def correlations_from_samples(beamformed_samples_1, beamformed_samples_2, output_sample_rate, slice_index_details):
+        """
+        Correlate two sets of beamformed samples together. Correlation matrices are used and
+        indices corresponding to lag pulse pairs are extracted.
+
+        :param      beamformed_samples_1:  The first beamformed samples.
+        :type       beamformed_samples_1:  ndarray [num_slices, num_beams, num_samples]
+        :param      beamformed_samples_2:  The second beamformed samples.
+        :type       beamformed_samples_2:  ndarray [num_slices, num_beams, num_samples]
+        :param      slice_index_details:   Details used to extract indices for each slice.
+        :type       slice_index_details:   list
+
+        :returns:   Correlations for slices.
+        :rtype:     list
+        """
+
+        # [num_slices, num_beams, num_samples]
+        # [num_slices, num_beams, num_samples]
+        correlated = np.einsum('ijk,ijl->ijkl', beamformed_samples_1,
+                               beamformed_samples_2.conj())
+
+        values = []
+        for s in slice_index_details:
+            if s['lags'].size == 0:
+                values.append(np.array([]))
+                continue
+            range_off = np.arange(s['num_range_gates'], dtype=np.int32) + s['first_range_off']
+
+            tau_in_samples = s['tau_spacing'] * 1e-6 * output_sample_rate
+
+            lag_pulses_as_samples = np.array(s['lags'], np.int32) * np.int32(tau_in_samples)
+
+            # [num_range_gates, 1, 1]
+            # [1, num_lags, 2]
+            samples_for_all_range_lags = (range_off[..., np.newaxis, np.newaxis] +
+                                          lag_pulses_as_samples[np.newaxis, :, :])
+
+            # [num_range_gates, num_lags, 2]
+            row = samples_for_all_range_lags[..., 1].astype(np.int32)
+
+            # [num_range_gates, num_lags, 2]
+            column = samples_for_all_range_lags[..., 0].astype(np.int32)
+
+            values_for_slice = correlated[s['slice_num'], :, row, column]
+
+            # [num_range_gates, num_lags, num_beams]
+            values_for_slice = np.einsum('ijk,j->kij', values_for_slice, s['lag_phase_offsets'])
+
+            values.append(values_for_slice)
+
+        return values
+
+
```

Comment:
I think that everything after this should get moved to a separate script in the testing module. No other python script has testing within it; this can be called from a separate utility script to test.